diffrax/_adjoint.py

```diff
@@ -852,3 +865,427 @@ def loop(
+                    "SDEs."
+                )
+            elif not isinstance(solver, AbstractStratonovichSolver):
+                warnings.warn(
+                    f"{solver.__class__.__name__} is not marked as converging to "
+                    "either the Itô or the Stratonovich solution. Note that "
+                    "`ReversibleAdjoint` will only produce the correct solution for "
+                    "Stratonovich SDEs."
+                )
+
+        solver = _Reversible(solver, self.l)
+        tprev = init_state.tprev
+        tnext = init_state.tnext
+        y = init_state.y
+
+        init_state = eqx.tree_at(
+            lambda s: s.solver_state,
+            init_state,
+            solver.init(terms, tprev, tnext, y, args),
+            is_leaf=_is_none,
+        )
+
+        init_state = eqx.tree_at(lambda s: s.y, init_state, object())
+        init_state = _nondiff_solver_controller_state(
+            self, init_state, passed_solver_state, passed_controller_state
+        )
+
+        final_state, aux_stats = _loop_reversible(
+            (y, args, terms),
+            self=self,
+            saveat=saveat,
+            init_state=init_state,
+            solver=solver,
+            event=event,
+            **kwargs,
+        )
+        final_state = _only_transpose_ys(final_state)
+        return final_state, aux_stats
+
+
+ReversibleAdjoint.__init__.__doc__ = r"""
+**Arguments:**
+
+- `l` - coupling parameter, defaults to `l=0.999`.
+
+The reversible solver introduces the coupled state $\{y_n, z_n\}_{n\geq 0}$ and the 
+coupling parameter $l\in (0, 1)$ mixes the states via $ly_n + (1-l)z_n$. This parameter 
+effects the stability of the reversible solver; decreasing it's value leads to greater 
+forward stability and increasing it's value leads to greater backward stability.
+
+In most cases the default value is sufficient. However, if you find yourself needing 
+greater control over stability it can be passed as an argument.
+"""
+
+_BaseSolverState = TypeVar("_BaseSolverState")
+_SolverState: TypeAlias = tuple[_BaseSolverState, Y]
+
+
+def _add_maybe_none(x, y):
+    if x is None:
+        return None
+    else:
+        return (ω(x) + ω(y)).ω
+
+
+class _Reversible(
+    AbstractAdaptiveSolver[_SolverState], AbstractWrappedSolver[_SolverState]
+):
+    """
+    Reversible solver method.
+
+    Allows any solver ([`diffrax.AbstractSolver`][]) to be made algebraically
+    reversible. This is a private API, exclusively for [`diffrax.ReversibleAdjoint`][].
+    """
+
+    solver: AbstractSolver
+    l: float = 0.999
+
+    @property
+    def term_structure(self):
+        return self.solver.term_structure
+
+    @property
+    def interpolation_cls(self):  # pyright: ignore
+        return self.solver.interpolation_cls
+
+    @property
+    def term_compatible_contr_kwargs(self):
+        return self.solver.term_compatible_contr_kwargs
```

Comment:
I think this, and `term_structure`, should not be necessary -- they are checked at the start of `diffeqsolve`, which we are already past. I think it may be neater to set these to `assert False`?